src/components/appointments/AppointmentFields.js

```diff
@@ -0,0 +1,189 @@
+        </select>
+      )}
+    </div>
+  );
+};
+
+const LocationField = ({value, onChange, url, urlQuery}) => {
+  const {loading, value: items} = useAsync(
+    async () => await requestOrEmpty(url, urlQuery, urlQuery.product_id),
+    [urlQuery.product_id]
+  );
+
+  return (
+    <div className={getBEMClassName('form-control')}>
+      <Label>
+        <FormattedMessage
+          description="Appointment location field label"
+          defaultMessage="Location"
+        />
+      </Label>
+
+      {loading ? (
+        <Loader />
+      ) : (
+        <select className={getBEMClassName('select')} value={value} onChange={onChange}>
+          <option value={''}></option>
+          {items.map(item => (
+            <option value={item.identifier} key={item.identifier}>
+              {item.name}
+            </option>
+          ))}
+        </select>
+      )}
+    </div>
+  );
+};
+
+const DateField = ({value, onChange, url, urlQuery}) => {
+  const intl = useIntl();
+  const {loading, value: items} = useAsync(
+    async () => await requestOrEmpty(url, urlQuery, urlQuery.product_id && urlQuery.location_id),
+    [urlQuery.product_id, urlQuery.location_id]
+  );
+
+  return (
+    <div className={getBEMClassName('form-control')}>
+      <Label>
+        <FormattedMessage description="Appointment date field label" defaultMessage="Date" />
+      </Label>
+
+      {loading ? (
+        <Loader />
+      ) : (
+        <select className={getBEMClassName('select')} value={value} onChange={onChange}>
+          <option value={''}></option>
+          {items.map(item => (
+            <option value={item.date} key={item.date}>
+              {getFormattedDateString(intl, item.date)}
+            </option>
+          ))}
+        </select>
+      )}
+    </div>
+  );
+};
+
+const TimeField = ({value, onChange, url, urlQuery}) => {
+  const intl = useIntl();
+  const {loading, value: items} = useAsync(
+    async () =>
+      await requestOrEmpty(
+        url,
+        urlQuery,
+        urlQuery.product_id && urlQuery.location_id && urlQuery.date
+      ),
+    [urlQuery.product_id, urlQuery.location_id, urlQuery.date]
+  );
+
+  return (
+    <div className={getBEMClassName('form-control')}>
+      <Label>
+        <FormattedMessage description="Appointment time field label" defaultMessage="Time" />
+      </Label>
+
+      {loading ? (
+        <Loader />
+      ) : (
+        <select className={getBEMClassName('select')} value={value} onChange={onChange}>
+          <option value={''}></option>
+          {items.map(item => (
+            <option value={item.time} key={item.time}>
+              {getFormattedTimeString(intl, item.time)}
+            </option>
+          ))}
+        </select>
+      )}
+    </div>
+  );
+};
+
+const AppointmentFields = () => {
+  const [product, setProduct] = useState('');
+  const [location, setLocation] = useState('');
+  const [date, setDate] = useState('');
+  const [time, setTime] = useState('');
```

Comment:
are you planning to check out Formik for form state management in a separate PR, or do we want this in here?

Reply:
refactored with formik